File: 695-max-area-of-island/max-area-of-island.py
class Solution:
    def dfs(self, grid: List[List[str]], i: int, j:int, visited:Set[Tuple[int]]) -> int:
        m = len(grid)
        n = len(grid[0])
        if i < 0 or i >= m or j < 0 or j >= n or (i,j) in visited or grid[i][j] == 0:
            return 0
        visited.add((i,j))
        score = 1
        score += self.dfs(grid, i+1, j, visited)
        score += self.dfs(grid, i, j+1, visited)
        score += self.dfs(grid, i-1, j, visited)
        score += self.dfs(grid, i, j-1, visited)
        return score


    def maxAreaOfIsland(self, grid: List[List[str]]) -> int:
        visited:Set[Tuple[int]] = set()
        m = len(grid)
        n = len(grid[0])
        solution = 0
        for i in range(m):
            for j in range(n):
                if (i,j) not in visited and grid[i][j] != 0:
                    solution = max(self.dfs(grid, i, j, visited), solution)
        return solution

File: 695-max-area-of-island/max-area-of-island.cpp
class Solution {
public:
    bool check_valid(vector<vector<int>>& grid, vector<vector<bool>>& visited, int x, int y) { 
        int n = grid.size();
        int m = grid[0].size();
        if(x < 0 || x >= n || y < 0 || y >= m) return false;
        if(grid[x][y] == 0) return false;
        if(visited[x][y]) return false;
        return true;
    }

    int maxAreaOfIsland(vector<vector<int>>& grid) {
        int n = grid.size();
        int m = grid[0].size();
        vector<vector<bool>> visited(n, vector<bool> (m, false));
        int ans = 0;
        for(int i = 0; i< n; i++){
            for(int j = 0; j < m; j++){
                if(check_valid(grid, visited, i, j)){
                    ans = max(ans, dfs(grid, visited, i, j));
                }
            }
        }
        return ans;
    }

    int dfs(vector<vector<int>>& grid, vector<vector<bool>>& visited, int x, int y){
        visited[x][y] = true;
        int size = 1;
        vector<pair<int, int>> directions({{1, 0}, {-1, 0}, {0, 1}, {0, -1}});
        for(auto& dir: directions){
            int new_x = x + dir.first;
            int new_y = y + dir.second;
            if(check_valid(grid, visited, new_x, new_y)){
                size += dfs(grid, visited, new_x, new_y);
            }
        }
        return size;
    }
};